conformance/tests/httproute-request-percentage-mirror.go

```diff
@@ -0,0 +1,236 @@
+	},
+	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+		var (
+			ns      = "gateway-conformance-infra"
+			routeNN = types.NamespacedName{Name: "request-percentage-mirror", Namespace: ns}
+			gwNN    = types.NamespacedName{Name: "same-namespace", Namespace: ns}
+			gwAddr  = kubernetes.GatewayAndHTTPRoutesMustBeAccepted(t, suite.Client, suite.TimeoutConfig, suite.ControllerName, kubernetes.NewGatewayRef(gwNN), routeNN)
+		)
+
+		kubernetes.HTTPRouteMustHaveResolvedRefsConditionsTrue(t, suite.Client, suite.TimeoutConfig, routeNN, gwNN)
+
+		testCases := []http.ExpectedResponse{
+			{
+				Request:   http.Request{Path: "/percent-mirror"},
+				Namespace: ns,
+				ExpectedRequest: &http.ExpectedRequest{
+					Request: http.Request{
+						Path: "/percent-mirror",
+					},
+				},
+				Backend: "infra-backend-v1",
+				MirroredTo: []http.MirroredBackend{
+					{
+						BackendRef: http.BackendRef{
+							Name:      "infra-backend-v2",
+							Namespace: ns,
+						},
+						Percent: ptrTo(int32(20)),
+					},
+				},
+			}, {
+				Request:   http.Request{Path: "/percent-mirror-fraction"},
+				Namespace: ns,
+				ExpectedRequest: &http.ExpectedRequest{
+					Request: http.Request{
+						Path: "/percent-mirror-fraction",
+					},
+				},
+				Backend: "infra-backend-v1",
+				MirroredTo: []http.MirroredBackend{
+					{
+						BackendRef: http.BackendRef{
+							Name:      "infra-backend-v2",
+							Namespace: ns,
+						},
+						Percent: ptrTo(int32(50)),
+					},
+				},
+			}, {
+				Request: http.Request{
+					Path: "/percent-mirror-and-modify-headers",
+					Headers: map[string]string{
+						"X-Header-Remove":     "remove-val",
+						"X-Header-Add-Append": "append-val-1",
+					},
+				},
+				ExpectedRequest: &http.ExpectedRequest{
+					Request: http.Request{
+						Path: "/percent-mirror-and-modify-headers",
+						Headers: map[string]string{
+							"X-Header-Add":        "header-val-1",
+							"X-Header-Add-Append": "append-val-1,header-val-2",
+							"X-Header-Set":        "set-overwrites-values",
+						},
+					},
+					AbsentHeaders: []string{"X-Header-Remove"},
+				},
+				Namespace: ns,
+				Backend:   "infra-backend-v1",
+				MirroredTo: []http.MirroredBackend{
+					{
+						BackendRef: http.BackendRef{
+							Name:      "infra-backend-v2",
+							Namespace: ns,
+						},
+						Percent: ptrTo(int32(35)),
+					},
+				},
+			},
+		}
+
+		for i := range testCases {
+			expected := testCases[i]
+			t.Run(expected.GetTestCaseName(i), func(t *testing.T) {
+				// Assert request succeeds before doing our distribution check
+				http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, expected)
+
+				// Override to not have more requests than expected
+				suite.TimeoutConfig.RequiredConsecutiveSuccesses = 1
+				// used to limit number of parallel go routines
+				semaphore := make(chan struct{}, concurrentRequests)
+				var wg sync.WaitGroup
```

Comment:
Why do we need to use concurrent clients here? Correct me If I'm wrong but what we want to check here is:
- send N requests
- check all the requests are received by the backend defined in the BackendRef
- check that n% of the requests are mirrored to the backend set in the mirror filter

Does it accurately summarize what we are trying to do here? If so, to me it's not clear why we do need to implement concurrent clients.

Reply:
since totalRequests are a high number of request, we are not sending them sequentially. Instead, we sending them in concurrently, BUT we limit the concurrency factor to be 10 (which is why the sempahore is used for).

Reply:
thanks for the explanation, sounds good to me